### ISSUE TITLE
Feature Request: Improve Pinniped OIDC Auth Experience with Browser Link Option

### ISSUE THREAD

**Issue by uozalp** [vmware-tanzu/pinniped#1868]
**Is your feature request related to a problem? Please describe.**

When using Pinniped OIDC auth with k9s, a link is printed in the middle of the screen prompting the user to log in. This link disrupts the user experience as it remains on the screen until k9s is closed and reopened.

**Describe the solution you'd like**

It would be beneficial to have an option to open the link in a browser without printing it to the screen. This would enhance usability and prevent disruption to the user's workflow.

**Are you considering submitting a PR for this feature?**

I'm open to contributing to this feature if guidance is provided by the maintainers.

- **How will this project improvement be tested?**
Testing would involve verifying that the link can be opened in a browser without being printed to the screen and ensuring that it does not disrupt the user's experience in k9s.

- **How does this change the current architecture?**
This change would likely involve modifying the logic responsible for printing the link to the screen.

- **How will this change be backwards compatible?**
This change should not affect existing functionality for users who prefer the current behavior. The option to print the link to the screen can remain as the default, with the new option providing an alternative.

- **How will this feature be documented?**
Documentation should include instructions on how to enable the option to open the link in a browser without printing it to the screen, along with any relevant configuration settings or commands.

**Additional context**

Adding this feature would improve the overall user experience when using Pinniped OIDC auth with k9s, making it more seamless and less disruptive.

https://github.com/vmware-tanzu/pinniped/blob/719cd75f64a198e8bd3b9f8ea8a29fcccb729e71/pkg/oidcclient/login.go#L648

![k9s](https://github.com/vmware-tanzu/pinniped/assets/54961032/a3d45473-e682-4253-959b-49ce07e0ca2a)

**Comment by cfryanr** [comment]
> We start our k9s sessions without passing the `--kubeconfig` parameter. Instead, we rely on HashiCorp Vault as our OIDC provider. Our `.kube/config` is configured accordingly, with certain data redacted for privacy reasons.

Thanks for the additional information. To confirm my understanding, I think what you are saying is that:
1. You do not need to use the `--kubeconfig` argument of k9s because your Pinniped-compatible kubeconfig is stored in your global kubeconfig file `$HOME/.kube/config`. So k9s will use your current context by default.
2. You are not using the Pinniped Supervisor. You are using the Pinniped Concierge with a `JWTAuthenticator` configured to point directly at some other OIDC provider (similar to the steps documented in https://pinniped.dev/docs/howto/concierge/configure-concierge-jwt/).

Is that right?

**Comment by uozalp** [comment]
That is correct.

**Comment by cfryanr** [comment]
Hi @uozalp,

I tried following the steps in https://pinniped.dev/docs/howto/concierge/configure-concierge-jwt/ and then using that kubeconfig with k9s. This should be similar to your setup.

When I first start k9s, it prompts me to log in before it displays the k9s UI. Are you seeing that behavior too? I'm using k9s v0.31.9, in case that matters.

So in your experience, is the login URL prompt displayed inside the k9s UI only after your tokens have expired? After you adjusted your tokens to live for 8 hours, do you only see it happen after running k9s for 8+ hours?

**Comment by uozalp** [comment]
We currently operate with three clusters. Upon initially launching k9s, we encounter the issue of text appearing in the center of the cluster screen. This also occurs when the authentication token expires or when switching contexts within k9s. While each browser session starts smoothly with proper authentication, the lingering text in the k9s interface remains a persistent challenge, necessitating app restarts. This issue affects all four members of our small team, comprising two Mac users and two Ubuntu users. We haven’t observed the behavior you’ve described on your end.

**Comment by uozalp** [comment]
We found a less-than-ideal workaround, resorting to calling `bash` instead of directly invoking `pinniped`. 

Introducing a silent option, something like `pinniped login oidc --enable-concierge --skip-console`, would greatly enhance the user experience.

```
    users:
    - name: ${var.environment_name}
      user:
        exec:
          apiVersion: client.authentication.k8s.io/v1beta1
          command: bash          
          args:
          - -c
          - pinniped login oidc --enable-concierge --concierge-api-group-suffix=pinniped.dev --concierge-authenticator-name=jwtauthenticator --concierge-authenticator-type=jwt --concierge-endpoint=https://${local.imp_proxy_external_endpoint} --concierge-ca-bundle-data=${base64encode(data.kubernetes_secret.this.data["ca.crt"])} --issuer=${vault_identity_oidc_provider.this.issuer} --client-id=${vault_identity_oidc_client.this.client_id} --scopes=openid,${vault_identity_oidc_scope.user.name},${vault_identity_oidc_scope.groups.name} --listen-port=12345 --request-audience=${vault_identity_oidc_client.this.client_id} 2>/dev/null
          env: []
          installHint: The pinniped CLI does not appear to be installed.  See https://get.pinniped.dev/cli
            for more details
          provideClusterInfo: true
```

**Comment by cfryanr** [comment]
I was not yet able to reproduce the initial login when k9s is first started interfering with the k9s UI, because k9s always invokes API calls that cause the login to happen before it launches the UI.

However, I was able to reproduce the case where, while using the k9s UI, my tokens have expired (including my refresh token) so I need to log in again, and the k9s UI is corrupted by Pinniped printing the login URL to stdout. I could also imagine how this same thing would happen when switching contexts in k9s (didn't try this myself).

Next I'll experiment with some different ideas for possible solutions. Certainly offering a new option could work. I'm also wondering if the Pinniped CLI could auto-detect this situation and avoid printing the login URL without needing a new option from the user.

**Comment by cfryanr** [comment]
By the way, is your OIDC provider returning a refresh token? If so, you will be able to see it in your `$HOME/.config/pinnniped/sessions.yaml` file while you have an active session going. If there is a refresh token, the Pinniped CLI will automatically use the refresh token to keep refreshing your ID tokens as they expire, until the refresh token stops working. Successful refreshes do not print anything to the screen. This would be an alternate way to make your sessions last longer. I wasn't sure if you meant that you configured your ID tokens or your refresh tokens to expire after 8 hours.

**Comment by uozalp** [comment]
My `sessions.yaml` looks like this. Is Id token a refresh token? 

```
apiVersion: config.supervisor.pinniped.dev/v1alpha1
kind: SessionCache
sessions:
- creationTimestamp: "2024-02-23T07:02:55Z"
  key:
    ...
  lastUsedTimestamp: "2024-02-23T07:02:55Z"
  tokens:
    access:
      expiryTimestamp: "2024-02-23T15:02:53Z"
      token: <TOKEN>
      type: Bearer
    id:
      claims:
        ...
      expiryTimestamp: "2024-02-23T15:02:53Z"
      token: <TOKEN>
```

**Comment by cfryanr** [comment]
If your OIDC provider had returned a refresh token then the sessions.yaml file would also include a `refresh` key, like:

```yaml
...
  tokens:
    access: ...
    id: ...
    refresh:
      token: <TOKEN>
...
```

Using a refresh token gives you the option to make the ID token lifetimes short and the refresh token lifetimes long. This might better a slightly better security posture because the ID token is sent to the cluster as a credential, as opposed to the refresh token which only lives on your computer and it only transmitted back to your OIDC provider (never to the cluster).

Exactly how to ask your OIDC provider to return a refresh token depends on the implementation of your OIDC provider. Sometimes you only need to add the `offline_access` scope to the scopes list when you run `pinniped get kubeconfig --oidc-scopes ...`. That's the most standard-compliant way for your OIDC provider to work, but unfortunately some providers work differently, so check the documentation for your provider to see if they support the `offline_access` scope or if they have some other means of requesting a refresh token on an authorize request.

**Comment by cfryanr** [comment]
By the way, an even better security posture would be to use the Pinniped Supervisor, because it takes care of a lot of these details for you, plus offers lots of other features to improve usability and security. Up to you to decide, of course. :)

**Comment by cfryanr** [comment]
Here is a draft solution: https://github.com/vmware-tanzu/pinniped/pull/1874

**Comment by cfryanr** [comment]
https://github.com/vmware-tanzu/pinniped/pull/1874 is merged. It will be included in the next release.

**Comment by cfryanr** [comment]
@uozalp Unfortunately, I don't think #1874 is working the way that I had hoped. I will follow up with another PR as soon as possible.

**Comment by cfryanr** [comment]
Trying a different approach in #1897.
